personhood/contract_popparty.go

```diff
@@ -0,0 +1,492 @@
+	var sigStrs []string
+	for _, sig := range signers {
+		sigStrs = append(sigStrs, sig.Identity().String())
+	}
+	signerCtrs, err := cl.GetSignerCounters(sigStrs...)
+	if err != nil {
+		return
+	}
+
+	fsBuf, err := protobuf.Encode(&fs)
+	if err != nil {
+		return
+	}
+	mr := make([]byte, 8)
+	binary.LittleEndian.PutUint64(mr, reward)
+	ctx := byzcoin.ClientTransaction{
+		Instructions: byzcoin.Instructions{byzcoin.Instruction{
+			InstanceID: byzcoin.NewInstanceID(dID),
+			Spawn: &byzcoin.Spawn{
+				ContractID: ContractPopPartyID,
+				Args: byzcoin.Arguments{{
+					Name:  "description",
+					Value: fsBuf,
+				}, {
+					Name:  "darcID",
+					Value: dID,
+				}, {
+					Name:  "miningReward",
+					Value: mr,
+				}},
+			},
+			SignerCounter: []uint64{signerCtrs.Counters[0] + 1},
+		}},
+	}
+	err = ctx.FillSignersAndSignWith(signers...)
+	if err != nil {
+		return
+	}
+	_, err = cl.AddTransactionAndWait(ctx, 5)
+	if err != nil {
+		return
+	}
+	return ctx.Instructions[0].DeriveID(""), nil
+}
+
+// PopPartyBarrier activates the barrier in the pop-party.
+func PopPartyBarrier(cl *byzcoin.Client, popIID byzcoin.InstanceID, signers ...darc.Signer) (err error) {
+	var sigStrs []string
+	for _, sig := range signers {
+		sigStrs = append(sigStrs, sig.Identity().String())
+	}
+	signerCtrs, err := cl.GetSignerCounters(sigStrs...)
+	if err != nil {
+		return
+	}
+
+	ctx := byzcoin.ClientTransaction{
+		Instructions: byzcoin.Instructions{byzcoin.Instruction{
+			InstanceID: popIID,
+			Invoke: &byzcoin.Invoke{
+				ContractID: ContractPopPartyID,
+				Command:    "barrier",
+			},
+			SignerCounter: []uint64{signerCtrs.Counters[0] + 1},
+		}},
+	}
+	err = ctx.FillSignersAndSignWith(signers...)
+	if err != nil {
+		return
+	}
+	_, err = cl.AddTransactionAndWait(ctx, 5)
+	return
+}
+
+// PopPartyFinalize sends the list of attendees to the party for finalization.
+func PopPartyFinalize(cl *byzcoin.Client, popIID byzcoin.InstanceID, atts Attendees, signers ...darc.Signer) (err error) {
+	var sigStrs []string
+	for _, sig := range signers {
+		sigStrs = append(sigStrs, sig.Identity().String())
+	}
+	signerCtrs, err := cl.GetSignerCounters(sigStrs...)
+	if err != nil {
+		return
+	}
+
+	attBuff, err := protobuf.Encode(&atts)
+	if err != nil {
+		return
+	}
+	ctx := byzcoin.ClientTransaction{
+		Instructions: byzcoin.Instructions{byzcoin.Instruction{
+			InstanceID: popIID,
+			Invoke: &byzcoin.Invoke{
+				ContractID: ContractPopPartyID,
+				Command:    "finalize",
+				Args: byzcoin.Arguments{
+					{
+						Name:  "attendees",
+						Value: attBuff,
+					},
+				},
+			},
+			SignerCounter: []uint64{signerCtrs.Counters[0] + 1},
+		}},
+	}
+	err = ctx.FillSignersAndSignWith(signers...)
+	if err != nil {
+		return
+	}
+	_, err = cl.AddTransactionAndWait(ctx, 5)
+	return
+}
+
+// PopPartyMine collects the reward. If 'atts' is nil, the party will be fetched from
+// byzcoin. One of 'coinIID' or 'd' must be set to either mine the coin, or spawn
```

Comment:
I'm not clear why this whole coinIID or DARC thing exists. I think it's because I've lost track of the big picture here. If I can't find that big picture doc someplace else in this PR, perhaps you can explain it better here?

Reply:
Updated comment:

```
// PopPartyMine is a method to be called by an outside client. It collects the reward for a given
// attendee of the party. For convenience, this can be called with some of the arguments being 'nil'.
// 
//   - atts - the list of the public keys of the attendees. If it is nil, the party will be fetched from
//     byzcoin. 
//   - coinIID - if set, 'd' must be nil. coinIID points to the coin InstanceID where the reward will be stored.
//   - d - if set, 'coinIID' must be nil. d is the darc that will be used to create a new coinInstance.
```